ArduCopter/mode_guided.cpp

```diff
@@ -436,29 +475,44 @@ void ModeGuided::pos_control_run()
     // set motors to full range
     motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
 
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    bool do_avoid = false;
+    guided_vel_target_cms = pos_control->get_vel_target_cms();
+#if AC_AVOID_ENABLED
+    // limit the velocity for obstacle/fence avoidance
+    copter.avoid.adjust_velocity(guided_vel_target_cms, pos_control->get_pos_xy_p().kP(), pos_control->get_max_accel_xy_cmss(), pos_control->get_pos_z_p().kP(), pos_control->get_max_accel_z_cmss(), G_Dt);
+    do_avoid = copter.avoid.limits_active();
+#endif
+
+    // send position and velocity targets to position controller
+    guided_accel_target_cmss.zero();
+    if (do_avoid) {
+        pos_control->input_vel_accel_xy(guided_vel_target_cms, guided_accel_target_cmss);
+        pos_control->input_vel_accel_z(guided_vel_target_cms, guided_accel_target_cmss, false);
+    } else {
+        guided_vel_target_cms.zero();
+        pos_control->input_pos_vel_accel_xyz(guided_pos_target_cm);
+    }
 
-    // WP_Nav has set the vertical position control targets
-    // run the vertical position controller and set output throttle
+    // run position controllers
+    pos_control->update_xy_controller();
     pos_control->update_z_controller();
 
     // call attitude controller
     if (auto_yaw.mode() == AUTO_YAW_HOLD) {
         // roll & pitch from waypoint controller, yaw rate from pilot
-        attitude_control->input_thrust_vector_rate_heading(wp_nav->get_thrust_vector(), target_yaw_rate);
+        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), target_yaw_rate);
     } else if (auto_yaw.mode() == AUTO_YAW_RATE) {
-        // roll & pitch from waypoint controller, yaw rate from mavlink command or mission item
-        attitude_control->input_thrust_vector_rate_heading(wp_nav->get_thrust_vector(), auto_yaw.rate_cds());
+        // roll & pitch from position-velocity controller, yaw rate from mavlink command or mission item
+        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), auto_yaw.rate_cds());
     } else {
         // roll, pitch from waypoint controller, yaw heading from GCS or auto_heading()
-        attitude_control->input_thrust_vector_heading(wp_nav->get_thrust_vector(), auto_yaw.yaw());
+        attitude_control->input_thrust_vector_heading(pos_control->get_thrust_vector(), auto_yaw.yaw());
     }
 }
 
-// guided_vel_control_run - runs the guided velocity controller
+// guided_velaccel_control_run - runs the guided velocity controller
```

Comment:
partially an existing issue but we should drop the "guided_" from the start of this comment.